### PR TITLE
Saml2 improvements

### PR DESCRIPTION
**Pull request by jzheaux**
Note that with the given changes then this sample is possible: https://github.com/jzheaux/spring-security-samples/tree/saml2-improvements/servlet/spring-boot/java/saml2/saml-extension-federation

### CODE REVIEW

**Review by marcusdacoregio — COMMENTED** [review]
Hey Josh, great work on this PR, it's a huge improvement for the SAML2 support.

I haven't noticed any major problems, just a few minor items and most of them are just questions that I had during the review.

The PR looks good to me.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/metadata/Saml2MetadataResponse.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.provider.service.metadata;
+
+public class Saml2MetadataResponse {
```
[the remark this answers is not in the document]
Comment (jzheaux) [reply saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/metadata/Saml2MetadataResponse.java:19]: I'm not up to date yet on the public API consequences of changing from a record to a class. For example, if we start out as a record but then later have something in here that requires a setter, can we change it to a class in a minor release?

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurerTests.java [hunk 2]
```diff
@@ -309,11 +307,7 @@ public void authenticationRequestWhenCustomAuthenticationRequestUriRepositoryThe
 
 	@Test
 	public void saml2LoginWhenLoginProcessingUrlWithoutRegistrationIdAndDefaultAuthenticationConverterThenValidates() {
-		assertThatExceptionOfType(BeanCreationException.class)
-				.isThrownBy(() -> this.spring.register(CustomLoginProcessingUrlDefaultAuthenticationConverter.class)
-						.autowire())
-				.havingRootCause().isInstanceOf(IllegalStateException.class)
-				.withMessage("loginProcessingUrl must contain {registrationId} path variable");
+		this.spring.register(CustomLoginProcessingUrlDefaultAuthenticationConverter.class).autowire();
```
[the remark this answers is not in the document]
Comment (jzheaux) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurerTests.java:310]: Actually, yes it is. :) Originally, it was validating that a url without `{registrationId}` would not be allowed, but now it is. Where startup used to fail, now it doesn't and this test confirms that.

That said, I can see how the name of the test might now be inaccurate. I'll adjust.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/OpenSamlAuthenticationTokenConverter.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	private Saml2AuthenticationToken tokenByEntityId(HttpServletRequest request) {
+		String serialized = request.getParameter(Saml2ParameterNames.SAML_RESPONSE);
+		String decoded = new String(samlDecode(serialized), StandardCharsets.UTF_8);
+		Response response = parse(decoded);
+		String issuer = response.getIssuer().getValue();
+		Iterator<RelyingPartyRegistration> registrations = this.registrations.findByAssertingPartyEntityId(issuer)
+				.iterator();
+		if (registrations.hasNext()) {
```
[the remark this answers is not in the document]
Comment (jzheaux) [reply saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/OpenSamlAuthenticationTokenConverter.java]: Changed the method to `findUniqueAssertingPartyEntityId` that returns `null` if there isn't a non-unique match.

**Pull request merged**